packages/utils/src/ast-utils/helpers.ts

```diff
@@ -43,33 +45,33 @@ export const isTokenOfTypeWithConditions = <
   conditions: Conditions,
 ): ((
   token: TSESTree.Token | null | undefined,
-) => token is TSESTree.Token & { type: TokenType } & Conditions) => {
+) => token is Extract<TSESTree.Token, { type: TokenType }> & Conditions) => {
   const entries = Object.entries(conditions) as ObjectEntries<TSESTree.Token>;
 
   return (
     token: TSESTree.Token | null | undefined,
-  ): token is TSESTree.Token & { type: TokenType } & Conditions =>
+  ): token is Extract<TSESTree.Token, { type: TokenType }> & Conditions =>
     token?.type === tokenType &&
-    entries.every(([key, value]) => token[key] === value);
+    entries.every(([key, value]) => key in token && token[key] === value);
```

Comment:
Quick eyeballing—`key` and `token` are both unions, so this `in` doesn't actually narrow either side, because TS won't do polynomial operations. Maybe `token[key as keyof TSESTree.Token>` would fix it? Not sure if there's a better way.

Reply:
Makes total sense, we need to make keys non-union to make it work

Also not sure if there's another way to do it

Reply:
Should be fixed now 👍